File: 538. 把二叉搜索树转换为累加树/Solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def traversal(self, root):
        if not root:
            return
        self.traversal(root.right)
        root.val += self.pre
        self.pre = root.val
        self.traversal(root.left)

    def convertBST(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        # 可以看出是从右向左计算累加的，也就是反中序遍历
        self.pre = 0 # 记录前一个节点的值
        self.traversal(root)
        return root

File: 538. 把二叉搜索树转换为累加树/Solution.cpp
#include<bits/stdc++.h>

using namespace std;


/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int pre = 0;
    TreeNode* convertBST(TreeNode* root) {
        traversal(root);
        return root;
    }
    void traversal(TreeNode* root){
        if(!root) return;
        traversal(root->right);
        root->val += pre;
        pre = root->val;
        traversal(root->left);
    }
};